pilot/pkg/config/kube/gateway/conversion.go

```diff
@@ -1449,8 +1450,24 @@ func convertGateways(r ConfigContext) ([]config.Config, map[parentKey]map[k8s.Se
 			if len(addressesToReport) == 0 {
 				// There are no external addresses, so report the internal ones
 				// TODO: should we always report both?
-				addressesToReport = internal
 				addrType = k8s.HostnameAddressType
+				for _, hostport := range internal {
+					svcname, _, _ := net.SplitHostPort(hostport)
+					svc := r.Context.ps.ServiceIndex.HostnameAndNamespace[host.Name(svcname)][obj.Namespace]
+					if svc.Attributes.ClusterExternalPorts != nil {
```

Comment:
@howardjohn Is the right way to check if the service is not LoadBalancer type?

Reply:
I am not sure I understand what we are trying to do here? ResolveGatewayInstances already partitions internal vs external addresses?

Reply:
It's trying to only use internal addresses (hostnames) if there are no external addresses AND the service in not LoadBalancer-type. This was to avoid the problem, you mentioned to me a couple of weeks ago, where we currently set the hostname initially, but then replace it after a delay, when the LoadBalancer address is ready. It's also stripping the port from the hostnames before setting the gateway addresses.

Reply:
I feel like we should have this logic inside ResolveGatewayInstances though and just change that as needed. Or at least use the same logic, which is not `ClusterExternalPorts` (although, probably ends up the same in practice)?

One thing I am not sure of  - using ClusterExternalPorts or ClusterExternalAddress may not handle the case where the LB is still `<provisioning>`

Reply:
Like we may not actually store the fact that "This is aLoadBalancer service but we don't have an IP yet" anywhere in model.service

Reply:
> Like we may not actually store the fact that "This is aLoadBalancer service but we don't have an IP yet" anywhere in model.service

Yes, that seems to be the case. If it's a LoadBalancer-type service that is still pending we will not have external addresses: https://github.com/istio/istio/blob/e8c809f026c18823bc57c1419b064b48158bb5ad/pilot/pkg/serviceregistry/kube/conversion.go#L141

I was hoping that external ports would be the differentiator, but maybe not?

Reply:
Looks like ClusterExternalPorts is just for NodePort?

Reply:
Ah right, I missed that - thought it was ClusterIP. So, I guess we have 2 options:

1. Add another field (e.g., LoadBalanceType bool).
2. Just strip the port and remove duplicates, but still set the hostname, for now.

IIRC, you mentioned that the conformance tests are soon going to be checking for gateway address to confirm ready, so if we go with option 2, we may fail the conformance test next time me update.

WDYT?

Reply:
(1) probably makes the most sense. Maybe am enum `Type`  under ServiceAttributes instead of a bool though?

Reply:
updated, PTAL